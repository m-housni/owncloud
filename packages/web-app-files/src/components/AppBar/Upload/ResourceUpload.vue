<template>
  <div>
    <oc-button :class="btnClass" justify-content="left" appearance="raw" @click="triggerUpload">
      <oc-resource-icon :resource="{ extension: '', isFolder }" size="medium" />
      <span :id="uploadLabelId">{{ buttonLabel }}</span>
    </oc-button>
    <input
      :id="inputId"
      ref="input"
      v-bind="inputAttrs"
      class="upload-input-target"
      type="file"
      :aria-labelledby="uploadLabelId"
      :name="isFolder ? 'file' : 'folder'"
      tabindex="-1"
    />
  </div>
</template>

<script>
export default {
  props: {
    btnLabel: {
      type: String,
      required: false,
      default: ''
    },
    btnClass: {
      type: String,
      required: false,
      default: ''
    },
    isFolder: {
      type: Boolean,
      required: false,
      default: false
    }
  },
  computed: {
    inputId() {
      if (this.isFolder) {
        return 'files-folder-upload-input'
      }
      return 'files-file-upload-input'
    },
    uploadLabelId() {
      if (this.isFolder) {
        return 'files-folder-upload-button'
      }
      return 'files-file-upload-button'
    },
    buttonLabel() {
      if (this.btnLabel) {
        return this.btnLabel
      }
      if (this.isFolder) {
        return this.$gettext('Folder')
      }
      return this.$gettext('Files')
    },
    inputAttrs() {
      if (this.isFolder) {
        return {
          webkitdirectory: true,
          mozdirectory: true,
          allowdirs: true
        }
      }
      return { multiple: true }
    }
  },
  mounted() {
    this.$uppyService.registerUploadInput(this.$refs.input)
  },
  beforeDestroy() {
    this.$uppyService.removeUploadInput(this.$refs.input)
  },
  methods: {
    triggerUpload() {
      this.$refs.input.click()
    }
  }
}
</script>

<style scoped>
.upload-input-target {
  position: absolute;
  left: -99999px;
}
</style>
