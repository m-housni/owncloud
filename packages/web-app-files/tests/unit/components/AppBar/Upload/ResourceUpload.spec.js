import { mount, createLocalVue } from '@vue/test-utils'
import ResourceUpload from '../../../../../src/components/AppBar/Upload/ResourceUpload.vue'
import DesignSystem from '@ownclouders/design-system'

const Translate = jest.fn()

describe('Resource Upload Component', () => {
  describe('file upload', () => {
    it('should render component', () => {
      const wrapper = mount(ResourceUpload, getOptions())
      expect(wrapper.exists()).toBeTruthy()
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('folder upload', () => {
    it('should render component', () => {
      const wrapper = mount(ResourceUpload, getOptions({ isFolder: true }))
      expect(wrapper.exists()).toBeTruthy()
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when upload file button is clicked', () => {
    it('should call "triggerUpload"', async () => {
      const wrapper = mount(ResourceUpload, {
        ...getOptions(),
        stubs: {
          'oc-icon': true,
          'oc-button': false
        }
      })

      const spyTriggerUpload = jest.spyOn(wrapper.vm, 'triggerUpload')
      const uploadButton = wrapper.find('button')
      const fileUploadInput = wrapper.find('#files-file-upload-input')
      fileUploadInput.element.click = jest.fn()
      await wrapper.vm.$forceUpdate()

      await uploadButton.trigger('click')

      expect(spyTriggerUpload).toHaveBeenCalledTimes(1)
      expect(fileUploadInput.element.click).toHaveBeenCalledTimes(1)
    })
  })
})

function getOptions(propsData = {}) {
  const localVue = createLocalVue()
  localVue.use(DesignSystem)
  return {
    propsData,
    localVue,
    directives: { Translate },
    mocks: {
      $uppyService: {
        registerUploadInput: jest.fn(),
        removeUploadInput: jest.fn()
      }
    }
  }
}
