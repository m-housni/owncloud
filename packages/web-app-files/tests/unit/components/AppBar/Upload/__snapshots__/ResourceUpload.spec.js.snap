// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resource Upload Component file upload should render component 1`] = `<div><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-left oc-button-gap-m oc-button-passive oc-button-passive-raw"><span class="oc-icon oc-icon-m oc-icon-passive oc-resource-icon oc-resource-icon-file"><!----></span> <span id="files-file-upload-button">Files</span></button> <input id="files-file-upload-input" type="file" aria-labelledby="files-file-upload-button" name="folder" tabindex="-1" multiple="multiple" class="upload-input-target"></div>`;

exports[`Resource Upload Component folder upload should render component 1`] = `<div><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-left oc-button-gap-m oc-button-passive oc-button-passive-raw"><span class="oc-icon oc-icon-m oc-icon-passive oc-resource-icon oc-resource-icon-folder"><!----></span> <span id="files-folder-upload-button">Folder</span></button> <input id="files-folder-upload-input" type="file" aria-labelledby="files-folder-upload-button" name="file" tabindex="-1" webkitdirectory="true" mozdirectory="true" allowdirs="true" class="upload-input-target"></div>`;
